src/org/opendatakit/briefcase/export/ExportToCsv.java

```diff
@@ -166,13 +110,86 @@ public static ExportOutcome export(FormDefinition formDef, ExportConfiguration c
     return exportOutcome;
   }
 
-  private static OutputStreamWriter getOutputStreamWriter(Path outputFile, Boolean overwrite, String header) {
-    if (Files.exists(outputFile) && !overwrite)
-      return newOutputStreamWriter(outputFile, APPEND);
-    // If we are not appending, open the file, truncate it if it already exists, and write the header
-    OutputStreamWriter osw = newOutputStreamWriter(outputFile, CREATE, TRUNCATE_EXISTING);
-    append(header, osw);
-    return osw;
+  // TODO Invert the relation between Submission and OutputFile to make OutputFile a mapper of Submission to avoid having to aggregate all lines into the List
+  static class OutputFile {
+    private static final OffsetDateTime MIN_SUBMISSION_DATE = OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, OffsetDateTime.now().getOffset());
+    private final String header;
+    private final Path output;
+    private final boolean overwrite;
+    private final List<Pair<OffsetDateTime, String>> linesBySubmissionDate = new ArrayList<>();
+    private final Function<Submission, Pair<OffsetDateTime, String>> submissionMapper;
+
+    OutputFile(String header, Path output, boolean overwrite, Function<Submission, Pair<OffsetDateTime, String>> submissionMapper) {
+      this.header = header;
+      this.output = output;
+      this.overwrite = overwrite;
+      this.submissionMapper = submissionMapper;
+    }
+
+    static OutputFile mainFile(FormDefinition formDefinition, ExportConfiguration configuration, boolean exportMedia) {
+      Path output = configuration.getExportDir()
+          .orElseThrow(BriefcaseException::new)
+          .resolve(configuration.getExportFileName().orElse(stripIllegalChars(formDefinition.getFormName()) + ".csv"));
+      Function<Submission, Pair<OffsetDateTime, String>> submissionMapper = submission -> Pair.of(
+          submission.getSubmissionDate().orElse(MIN_SUBMISSION_DATE),
+          getMainSubmissionLines(
+              submission,
+              formDefinition.getModel(),
+              formDefinition.isFileEncryptedForm(),
+              exportMedia,
+              configuration.getExportMediaPath()
+          )
+      );
+      return new OutputFile(
+          getMainHeader(formDefinition.getModel(), formDefinition.isFileEncryptedForm()),
+          output,
+          configuration.getOverwriteExistingFiles().orElse(false),
+          submissionMapper
+      );
+    }
+
```

Comment:
I’m curious why you prefer the separate Function object (left) to the in-line form (right).

![screen shot 2018-06-30 at 4 02 03 pm](https://user-images.githubusercontent.com/55099/42129634-1402130a-7c7f-11e8-9d2f-847d4c69aa55.png)